### PR TITLE
Documentation for ext-mongodb 1.16.0

### PR DESCRIPTION
**Pull request by alcaeus**

### CODE REVIEW

**Comment by alcaeus** [comment]
This is ready for review @jmikola

File: reference/mongodb/bson.xml
Comment (jmikola) [inline reference/mongodb/bson.xml]: I'll skip reviewing the MongoDB\BSON\Value docs based on our conversation earlier today.

File: reference/mongodb/bson/document.xml
Comment (jmikola) [inline reference/mongodb/bson/document.xml]:
```suggestion
    and cannot be modified.
```

"can not" appears 75 times across the docs, while "cannot" has 500+ occurences.

File: reference/mongodb/bson/document/construct.xml
Comment (jmikola) [inline reference/mongodb/bson/document/construct.xml]:
```suggestion
   static factory methods and cannot be instantiated directly.
```

File: reference/mongodb/bson/document/construct.xml
Comment (jmikola) [inline reference/mongodb/bson/document/construct.xml:33]: I assume you included the same reference on all methods. In this case, would it make more sense to reference the three static factory methods?

File: reference/mongodb/bson/document/frombson.xml
Comment (jmikola) [inline reference/mongodb/bson/document/frombson.xml]:
```suggestion
   <member>Throws <classname>MongoDB\Driver\Exception\UnexpectedValueException</classname> if <parameter>bson</parameter> is an invalid BSON string or contains more than a single document.</member>
```

File: reference/mongodb/bson/document/fromjson.xml
Comment (jmikola) [inline reference/mongodb/bson/document/fromjson.xml]: Consider the docs for [`fromJSON()`](https://www.php.net/manual/en/function.mongodb.bson-fromjson.php):

> Converts an [» extended JSON](https://www.mongodb.com/docs/manual/reference/mongodb-extended-json/) string to its BSON representation.

Mentioning Extended JSON beyond just a "see also" link would be helpful.

File: reference/mongodb/bson/document/fromphp.xml
Comment (jmikola) [inline reference/mongodb/bson/document/fromphp.xml]:
```suggestion
      numeric keys, the numeric values are converted to strings and used as
```

File: reference/mongodb/bson/document/get.xml
Comment (jmikola) [inline reference/mongodb/bson/document/get.xml]: I expect this needs to be changed to `mixed`, so I'll hold off on reviewing this file.

File: reference/mongodb/bson/document/getiterator.xml
Comment (jmikola) [inline reference/mongodb/bson/document/getiterator.xml]: This will need to be changed.

File: reference/mongodb/bson/document/serialize.xml
Comment (jmikola) [inline reference/mongodb/bson/document/serialize.xml:6]: I realize this is beyond the scope of the PR, but I remembered that we don't document the `__serialize()` and `_unserialize()` magic methods, just the [deprecated Serializable interface](https://php.watch/versions/8.1/serializable-deprecated).

Do you think we should do both? I noticed some other extensions (e.g. [GMP](https://www.php.net/manual/en/gmp.serialize.php)) document their magic methods. As-is, users might not even realize we implement those at all. If you concur, let's open and backlog a PHPC ticket to do so.

Offhand, I'm not sure how the filename would work, as the other extensions don't use underscores.

File: reference/mongodb/bson/document/tocanonicalextendedjson.xml
Comment (jmikola) [inline reference/mongodb/bson/document/tocanonicalextendedjson.xml:15]: Consider pulling in some content from the description of [`toCanonicalExtendedJSON()`](https://www.php.net/manual/en/function.mongodb.bson-tocanonicalextendedjson.php).

File: reference/mongodb/bson/document/tophp.xml
Comment (jmikola) [inline reference/mongodb/bson/document/tophp.xml:15]: Pull in content from [`toPHP()`](https://www.php.net/manual/en/function.mongodb.bson-tophp.php).

File: reference/mongodb/bson/document/torelaxedextendedjson.xml
Comment (jmikola) [inline reference/mongodb/bson/document/torelaxedextendedjson.xml:15]: Pull in content from [`toRelaxedExtendedJSON()`](https://www.php.net/manual/en/function.mongodb.bson-torelaxedextendedjson.php).

File: reference/mongodb/bson/int64.xml
Comment (jmikola) [inline reference/mongodb/bson/int64.xml]:
```suggestion
    is used when a 64-bit integer cannot be represented as a PHP integer on
    32-bit platforms. These objects support overloaded
```

Grammar nit, but I wouldn't resort to using "i.e." here unless we were explaining the full situation of a 64-bit integer _type_ containing a value beyond the range of a signed 32-bit integer.

File: reference/mongodb/bson/int64.xml
Comment (jmikola) [inline reference/mongodb/bson/int64.xml]: This needs to be rewritten to note that Document and PackedArray will simply return Int64 for all 64-bit BSON integer type.

File: reference/mongodb/bson/int64.xml
Comment (jmikola) [inline reference/mongodb/bson/int64.xml]:
```suggestion
         overloaded arithmetic, bitwise, and comparison operators.
```

File: reference/mongodb/bson/int64/construct.xml
Comment (jmikola) [inline reference/mongodb/bson/int64/construct.xml]:
```suggestion
   <methodparam><type class="union"><type>int</type><type>string</type></type><parameter>value</parameter></methodparam>
```

Not sure if this was copied from something else, but the stub and actual implementation is:

```
final public function __construct(int|string $value)
```

File: reference/mongodb/bson/int64/construct.xml
Comment (jmikola) [inline reference/mongodb/bson/int64/construct.xml:36]: This is missing an errors section:

```xml
<refsect1 role="errors">
 &reftitle.errors;
 <simplelist>
  &mongodb.throws.argumentparsing;
  <member>Throws <classname>MongoDB\Driver\Exception\InvalidArgumentException</classname> if string <parameter>value</parameter> cannot be parsed as a 64-bit integer.</member>
 </simplelist>
</refsect1>
```

There's also an exception thrown for an invalid type, but I think that's covered by `&mongodb.throws.argumentparsing;`.

I'll defer to you if you want to add an examples section or not. If so, we should demonstrate construction from both an int and string.

File: reference/mongodb/bson/iterator/construct.xml
Comment (jmikola) [inline reference/mongodb/bson/iterator/construct.xml]:
```suggestion
   <methodname>MongoDB\BSON\PackedArray::getIterator</methodname> and cannot be
```

File: reference/mongodb/bson/iterator/current.xml
Comment (jmikola) [inline reference/mongodb/bson/iterator/current.xml]: This will need to be updated.

File: reference/mongodb/bson/iterator/key.xml
Comment (jmikola) [inline reference/mongodb/bson/iterator/key.xml]:
```suggestion
   key will be an <type>int</type>.
```

File: reference/mongodb/bson/iterator/key.xml
Comment (jmikola) [inline reference/mongodb/bson/iterator/key.xml:27]: Does it make sense to refer to the class names here? I'm fine with this as-is -- just wondering if this language was intentional.

File: reference/mongodb/bson/iterator/next.xml
Comment (jmikola) [inline reference/mongodb/bson/iterator/next.xml:13]: I realize the return type is `void`, but when checking the implementation I realized it unconditionally increments `intern->key`:

```c
intern->valid = bson_iter_next(&intern->iter);
intern->key++;
php_phongo_iterator_free_current(intern);
```

Does it make sense to do so if the position is not valid? I suppose once we reach an invalid position, we'd never get back to a valid position where `Iterator::key()` could be called successfully -- so maybe this isn't a concern.

File: reference/mongodb/bson/iterator/valid.xml
Comment (jmikola) [inline reference/mongodb/bson/iterator/valid.xml]: Consider replacing "iterator" with one of the following:

 * "current position"
 * "current iterator position"
 * "iterator's current position"

There is prior art for the first two bullets in other `Iterator::valid()` implementations within the PHP docs.

File: reference/mongodb/bson/packedarray/construct.xml
Comment (jmikola) [inline reference/mongodb/bson/packedarray/construct.xml]:
```suggestion
   static factory methods and cannot be instantiated directly.
```

File: reference/mongodb/bson/packedarray/get.xml
Comment (jmikola) [inline reference/mongodb/bson/packedarray/get.xml]: This will need to change change.

File: reference/mongodb/bson/packedarray/get.xml
Comment (jmikola) [inline reference/mongodb/bson/packedarray/get.xml]:
```suggestion
      The index to retrieve from the array.
```

Parameter descriptions typically end with a period. Feel free to double-check in case I missed others.

File: reference/mongodb/bson/packedarray/getiterator.xml
Comment (jmikola) [inline reference/mongodb/bson/packedarray/getiterator.xml]: This will need to change.

File: reference/mongodb/bson/value.xml
Comment (jmikola) [inline reference/mongodb/bson/value.xml]: Skipping all of the Value classes, since I expect these will be removed.

File: reference/mongodb/bson/document/fromjson.xml
Comment (alcaeus) [reply reference/mongodb/bson/document/fromjson.xml]: Good point. I've used the language from `fromJSON`.

File: reference/mongodb/bson/document/serialize.xml
Comment (alcaeus) [reply reference/mongodb/bson/document/serialize.xml:6]: Created PHPC-2247 to track this. The issue of filenames is an interesting one, as the `__toString` documentation is contained in `tostring.xml`.

File: reference/mongodb/bson/document/construct.xml
Comment (alcaeus) [reply reference/mongodb/bson/document/construct.xml:33]: I've added links to the factories and cross-linked them in the factory documentation as well.

File: reference/mongodb/bson/int64/construct.xml
Comment (alcaeus) [reply reference/mongodb/bson/int64/construct.xml]: Definitely some copy-pasta left over. Fixed now.

File: reference/mongodb/bson/int64/construct.xml
Comment (alcaeus) [reply reference/mongodb/bson/int64/construct.xml:36]: I skipped the examples section for now.

File: reference/mongodb/bson/iterator/key.xml
Comment (alcaeus) [reply reference/mongodb/bson/iterator/key.xml:27]: We could use the class names here, but that's just the origin of the cursor. The idea was to differentiate the language based on the kind of structure the iterator is iterating over, not the origin of the structure (e.g. a `BSON\Document` instance).

File: reference/mongodb/bson/iterator/next.xml
Comment (alcaeus) [reply reference/mongodb/bson/iterator/next.xml:13]: Interesting find, and a good point. This could be a problem if we don't properly handle rewinding the iterator, but then `next` might as well guard against an invalid iterator and no-op if the iterator is already invalid. While that strategy could then cause unexpected behaviour if the underlying `bson_t` were to change (which currently isn't possible), this is already an issue in libbson and explained in the `bson_iter_t` docs:
> The [bson_t](http://mongoc.org/libbson/current/bson_t.html) MUST be valid for the lifetime of the iter and it is an error to modify the [bson_t](http://mongoc.org/libbson/current/bson_t.html) while using the iter.

I'll add some tests in https://github.com/mongodb/mongo-php-driver/pull/1435.

File: reference/mongodb/bson/iterator/valid.xml
Comment (alcaeus) [reply reference/mongodb/bson/iterator/valid.xml]: Replaced with "current iterator position".

File: reference/mongodb/bson/packedarray/get.xml
Comment (alcaeus) [reply reference/mongodb/bson/packedarray/get.xml]: Went through and fixed a couple more occurrences in the new docs.

**Comment by alcaeus** [comment]
PR updated with `MongoDB\BSON\Value` dropped and return types for classes changed. I also added language to highlight the return type for int64 types in the following methods:
* `MongoDB\BSON\Document::get`
* `MongoDB\BSON\Document::toPHP`
* `MongoDB\BSON\Iterator::current`
* `MongoDB\BSON\PackedArray::get`
* `MongoDB\BSON\PackedArray::toPHP`

File: reference/mongodb/bson/document/serialize.xml
Comment (jmikola) [reply reference/mongodb/bson/document/serialize.xml:6]: > The issue of filenames is an interesting one

I left a comment about that in [PHPC-2247](https://jira.mongodb.org/browse/PHPC-2247?focusedCommentId=5505599&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-5505599).

**Review by jmikola — APPROVED** [review]

**Comment by alcaeus** [comment]
Merged manually as https://github.com/php/doc-en/commit/52c3812df42c03188138c6930e49ecd7188f0e86

**Pull request closed without merging**
